File: frontend/src/models/recommendation/RecommendationFactory.ts
import Recommendation from "./Recommendation";
import {RecommendationType} from "./RecommendationTypes";
import RecommendMatrixSciPy from "./RecommendMatrixSciPy";

export abstract class RecommendationFactory {
    static getRecommendation(recommendation: any) : Recommendation {
        switch (recommendation.type) {
            case RecommendationType.RECOMMEND_MATRIX_SCIPY:
            case RecommendationType.RECOMMEND_MATRIX_CLASS_VECTORIZATION:
            case RecommendationType.RECOMMEND_MATRIX_ENTITY_VECTORIZATION:
            case RecommendationType.RECOMMEND_MATRIX_FUNCTIONALITY_VECTORIZATION_CALLGRAPH:
            case RecommendationType.RECOMMEND_MATRIX_FUNCTIONALITY_VECTORIZATION_SEQUENCE_ACCESSES:
                return new RecommendMatrixSciPy(recommendation);
            default:
                throw new Error('Type ' + recommendation.type + ' unknown.');
        }
    }
}

File: backend/src/main/java/pt/ist/socialsoftware/mono2micro/recommendation/domain/RecommendationFactory.java
package pt.ist.socialsoftware.mono2micro.recommendation.domain;

import pt.ist.socialsoftware.mono2micro.recommendation.dto.*;
import pt.ist.socialsoftware.mono2micro.strategy.domain.Strategy;

import static pt.ist.socialsoftware.mono2micro.recommendation.domain.RecommendationsType.*;

public class RecommendationFactory {
    public static Recommendation getRecommendation(RecommendationDto recommendationDto) {
        switch (recommendationDto.getType()) {
            case RECOMMEND_MATRIX_SCIPY:
            case RECOMMEND_MATRIX_CLASS_VECTORIZATION:
            case RECOMMEND_MATRIX_ENTITY_VECTORIZATION:
            case RECOMMEND_MATRIX_FUNCTIONALITY_VECTORIZATION_CALLGRAPH:
            case RECOMMEND_MATRIX_FUNCTIONALITY_VECTORIZATION_SEQUENCE_ACCESSES:
                return new RecommendMatrixSciPy((RecommendMatrixSciPyDto) recommendationDto);
            default:
                throw new RuntimeException("The type \"" + recommendationDto.getType() + "\" is not a valid decomposition type for recommendation creation.");
        }
    }

    public static Recommendation getRecommendation(Strategy strategy, RecommendationDto recommendationDto) {
        Recommendation recommendation = getRecommendation(recommendationDto);
        recommendation.setCompleted(false);

        // Get new name
        int i = 0;
        String recommendationName;
        do {
            recommendationName = strategy.getName() + " | " + " Recommendation " + ++i;
        } while (strategy.containsRecommendationName(recommendationName));
        recommendation.setName(recommendationName);
        recommendation.setRecommendationResultName(recommendation.getName() + "_recommendationResult");
        strategy.addRecommendation(recommendation);
        recommendation.setStrategy(strategy);

        return recommendation;
    }
}
